pkg/controllers/disruption/types.go

```diff
@@ -60,60 +62,31 @@ type Candidate struct {
-	}
-	// skip any candidates that are already marked for deletion and being handled
-	if node.MarkedForDeletion() {
-		return nil, fmt.Errorf("state node is marked for deletion")
-	}
-	// skip candidates that aren't initialized
-	if !node.Initialized() {
-		return nil, fmt.Errorf("state node isn't initialized")
+	if err := node.IsDisruptable(ctx, kubeClient); err != nil {
+		recorder.Publish(disruptionevents.Blocked(node.Node, node.NodeClaim, err.Error())...)
+		return nil, err
 	}
 	// If the orchestration queue is already considering a candidate we want to disrupt, don't consider it a candidate.
 	if queue.HasAny(node.ProviderID()) {
-		return nil, fmt.Errorf("candidate is already being deprovisioned")
-	}
-	if _, ok := node.Annotations()[v1beta1.DoNotDisruptAnnotationKey]; ok {
-		recorder.Publish(disruptionevents.Blocked(node.Node, node.NodeClaim, fmt.Sprintf("Disruption is blocked with the %q annotation", v1beta1.DoNotDisruptAnnotationKey))...)
-		return nil, fmt.Errorf("disruption is blocked through the %q annotation", v1beta1.DoNotDisruptAnnotationKey)
-	}
-	// check whether the node has all the labels we need
-	for _, label := range []string{
-		v1beta1.CapacityTypeLabelKey,
-		v1.LabelTopologyZone,
-	} {
-		if _, ok := node.Labels()[label]; !ok {
-			recorder.Publish(disruptionevents.Blocked(node.Node, node.NodeClaim, fmt.Sprintf("Required label %q doesn't exist", label))...)
-			return nil, fmt.Errorf("state node doesn't have required label %q", label)
-		}
-	}
-	nodePoolName, ok := node.Labels()[v1beta1.NodePoolLabelKey]
-	if !ok {
-		return nil, fmt.Errorf("state node doesn't have the Karpenter owner label")
+		return nil, fmt.Errorf("candidate is already being disrupted")
 	}
+	// We know that the node will have the label key because of the node.IsDisruptable check above
+	nodePoolName := node.Labels()[v1beta1.NodePoolLabelKey]
 	nodePool := nodePoolMap[nodePoolName]
 	instanceTypeMap := nodePoolToInstanceTypesMap[nodePoolName]
 	// skip any candidates where we can't determine the nodePool
 	if nodePool == nil || instanceTypeMap == nil {
-		recorder.Publish(disruptionevents.Blocked(node.Node, node.NodeClaim, fmt.Sprintf("Owning nodepool %q not found", nodePoolName))...)
+		recorder.Publish(disruptionevents.Blocked(node.Node, node.NodeClaim, fmt.Sprintf("NodePool %q not found", nodePoolName))...)
 		return nil, fmt.Errorf("nodepool %q can't be resolved for state node", nodePoolName)
 	}
 	instanceType := instanceTypeMap[node.Labels()[v1.LabelInstanceTypeStable]]
 	// skip any candidates that we can't determine the instance of
 	if instanceType == nil {
-		recorder.Publish(disruptionevents.Blocked(node.Node, node.NodeClaim, fmt.Sprintf("Instance type %q not found", node.Labels()[v1.LabelInstanceTypeStable]))...)
+		recorder.Publish(disruptionevents.Blocked(node.Node, node.NodeClaim, fmt.Sprintf("Instance Type %q not found", node.Labels()[v1.LabelInstanceTypeStable]))...)
```

Comment:
Why do we think that some of these checks belong at the top-level codepath and others belong in statenode. Is there a way to push everything into a common place or are there specific limitations here that we have to consider?